aidial_analytics_realtime/app.py

```
@@ -298,28 +302,76 @@ async def on_log_messages(
-            statuses.append({"status": "error", "error": str(e)})
-        else:
-            statuses.append({"status": "success"})
+    n = len(data)
+    logger.info(f"number of messages: {n}")
+
+    statuses: list[dict] = []
+
+    async with Timer(logger.debug):
+
+        async def _task(i: int, message_str: str) -> dict:
+            add_logger_prefix(f"[{i}/{n}]")
+
+            async with Timer(logger.debug):
+                return await process_message(
+                    json.loads(message_str),
+                    influx_writer,
+                    topic_model,
+                    rates_calculator,
+                )
+
+        statuses = await asyncio.gather(
+            *[
+                _task(i, message["message"])
+                for i, message in enumerate(data, start=1)
+            ]
+        )
+
+    if logger.isEnabledFor(logging.DEBUG):
+        logger.debug(f"response: {json.dumps(statuses)}")
 
     # Returning 200 code even if processing of some messages has failed,
     # since the log broker that sends the messages may decide to retry the failed requests.
     return JSONResponse(content=statuses, status_code=200)
 
 
+async def process_message(
+    message: dict,
+    influx_writer: InfluxWriterAsync,
+    topic_model: TopicModel,
+    rates_calculator: RatesCalculator,
+) -> dict:
+    def _error(reason: str | None = None) -> dict:
+        error = str(sys.exc_info()[1])
+        ret = {"status": "error"}
+        if error:
+            ret["error"] = error
+        if reason:
+            ret["reason"] = reason
+            logger.error(reason)
+        else:
+            logger.exception("caught exception")
+        return ret
+
+    try:
+        await on_log_message(
+            message, influx_writer, topic_model, rates_calculator
+        )
+        logger.info("success")
+        return {"status": "success"}
+    except starlette.requests.ClientDisconnect:
+        return _error("client disconnect")
```

Why not just use `starlette.requests.ClientDisconnect` here?
I'm not sure that all this extra code (including defining a function within a function) worth it since the error is not returned to the response anyway.

---

It's mainly to see the errors in the logs. 
The typical errors are reported briefly (as "timeout" or "client disconnect").
The rest of the unexpected exceptions are reported fully with a stack trace.

Previously the errors were completely silent.